### ISSUE TITLE
cross compile vim: --with-tlib

### ISSUE THREAD

**Issue by companion-vu** [vim/vim#2058]
I want to cross compile vim on Ubuntu 16.04 LTS for armv7. When I try to configure, I got the error:

    …
    checking for tgetent in -lcurses... no
    no terminal library found
    checking for tgetent()... configure: error: NOT FOUND!
          You need to install a terminal library; for example ncurses.
          Or specify the name of the library with --with-tlib.

Therefore, I cross compiled ncurses and I install ncurses for arm in *~/work/ncurses/install*:

    #~/work/ncurses/install# l
    insgesamt 12K
    drwxrwxr-x 2 root root 4,0K Sep  3 16:56 bin
    drwxrwxr-x 2 root root 4,0K Sep  3 16:56 lib
    drwxrwxr-x 5 root root 4,0K Sep  3 16:56 usr
    #~/work/ncurses/install# cd bin
    #~/work/ncurses/install/bin# l
    insgesamt 192K
    lrwxrwxrwx 1 root root   23 Sep  3 16:56 arm-linux-gnueabihf-captoinfo -> arm-linux-gnueabihf-tic
    -rwxr-xr-x 1 root root 8,3K Sep  3 16:56 arm-linux-gnueabihf-clear
    -rwxr-xr-x 1 root root  47K Sep  3 16:56 arm-linux-gnueabihf-infocmp
    lrwxrwxrwx 1 root root   23 Sep  3 16:56 arm-linux-gnueabihf-infotocap -> arm-linux-gnueabihf-tic
    lrwxrwxrwx 1 root root   24 Sep  3 16:56 arm-linux-gnueabihf-reset -> arm-linux-gnueabihf-tset
    -rwxr-xr-x 1 root root  14K Sep  3 16:56 arm-linux-gnueabihf-tabs
    -rwxr-xr-x 1 root root  51K Sep  3 16:56 arm-linux-gnueabihf-tic
    -rwxr-xr-x 1 root root  14K Sep  3 16:56 arm-linux-gnueabihf-toe
    -rwxr-xr-x 1 root root  14K Sep  3 16:56 arm-linux-gnueabihf-tput
    -rwxr-xr-x 1 root root  20K Sep  3 16:56 arm-linux-gnueabihf-tset
    -rwxr-xr-x 1 root root 5,2K Sep  3 16:56 ncurses5-config
    #~/work/ncurses/install/bin# l ../lib
    insgesamt 1004K
    lrwxrwxrwx 1 root root   12 Sep  3 16:56 libcurses.a -> libncurses.a
    lrwxrwxrwx 1 root root   17 Sep  3 16:56 libcurses.so -> libncurses.so.5.9
    -rw-r--r-- 1 root root  79K Sep  3 16:56 libform.a
    lrwxrwxrwx 1 root root   12 Sep  3 16:56 libform.so -> libform.so.5
    lrwxrwxrwx 1 root root   14 Sep  3 16:56 libform.so.5 -> libform.so.5.9
    -rwxrwxr-x 1 root root  49K Sep  3 16:56 libform.so.5.9
    -rw-r--r-- 1 root root  44K Sep  3 16:56 libmenu.a
    lrwxrwxrwx 1 root root   12 Sep  3 16:56 libmenu.so -> libmenu.so.5
    lrwxrwxrwx 1 root root   14 Sep  3 16:56 libmenu.so.5 -> libmenu.so.5.9
    -rwxrwxr-x 1 root root  30K Sep  3 16:56 libmenu.so.5.9
    -rw-r--r-- 1 root root 377K Sep  3 16:56 libncurses.a
    -rw-r--r-- 1 root root 128K Sep  3 16:56 libncurses++.a
    lrwxrwxrwx 1 root root   15 Sep  3 16:56 libncurses.so -> libncurses.so.5
    lrwxrwxrwx 1 root root   17 Sep  3 16:56 libncurses.so.5 -> libncurses.so.5.9
    -rwxrwxr-x 1 root root 230K Sep  3 16:56 libncurses.so.5.9
    -rw-r--r-- 1 root root  19K Sep  3 16:56 libpanel.a
    lrwxrwxrwx 1 root root   13 Sep  3 16:56 libpanel.so -> libpanel.so.5
    lrwxrwxrwx 1 root root   15 Sep  3 16:56 libpanel.so.5 -> libpanel.so.5.9
    -rwxrwxr-x 1 root root  13K Sep  3 16:56 libpanel.so.5.9
    #~/work/ncurses/install/bin# l ../usr
    insgesamt 12K
    drwxrwxr-x 2 root root 4,0K Sep  3 16:56 include
    drwxrwxr-x 6 root root 4,0K Sep  3 16:56 man
    drwxrwxr-x 4 root root 4,0K Sep  3 16:56 share

I try now to configure with *--with-tlib=~/work/ncurses/install/lib* but I get the error:

    …
    checking for stack_t... yes
    checking whether stack_t has an ss_base field... no
    checking --with-tlib argument... ~/work/ncurses/install/lib
    checking for linking with ~/work/ncurses/install/lib library... configure: error: FAILED

How can I make it work?
Thanks in advance.

**Comment by nuko8** [comment]
> How can I make it work?

Although I'm not sure if the mentioned cross compile is possible, I can give a comment upon `--with-tlib` usage.

From `src/configure.ac:3300`, it follows that what you need to specify with `--with-tlib=` is *the name of the library*,  as suggested by the first error message in your report.  That means, for your case, you are supposed to set it to either `ncurses` or `curses`.   Also, the configure script needs to know where the terminal library to use is.  You can tell the script the location of it via the `LDFLAGS` environment variable.  Accordingly, as far as terminal libraries are concerned, the script needs to run like
```
LDFLAGS=/work/ncurses/install/lib ./configure --with-tlib=ncurses
```
(Probably, you don't need to set `--with-tlib` explicitly if the terminal libraries you have there are only `ncurses` and `curses`; if this is the case, the script automatically selects `ncurses`.)

**Comment by companion-vu** [comment]
Hi nuko8,
Thank for your answer. I am right now await from my box but I run it like: 

    export LDFLAGS=~/work/ncurses/install/lib
    LDFLAGS=-L~/work/ncurses/install/lib ../src/configure --target=$TARGET --prefix=$PREFIX --with-sysroot=$TARGET --with-arch=$TARGET CC=arm-linux-gnueabihf-gcc --host=$TARGET --with-tlib=ncurses

but I got an error something like lib not found or so. I’ll try to leave out –with-tlib and come back.

**Comment by k-takata** [comment]
How about using `$HOME` instead of a tilde?
```
LDFLAGS=-L$HOME/work/ncurses/install/lib ./configure
```

**Comment by companion-vu** [comment]
This works now:

    export TARGET=arm-linux-gnueabihf
    export vim_cv_toupper_broken="set"
    export LDFLAGS=
    export PREFIX=~/work/vim/build
    echo $TARGET $vim_cv_toupper_broken $LDFLAGS $PREFIX
    ../src/configure --target=$TARGET --prefix=$PREFIX --with-sysroot=$TARGET --with-arch=$TARGET CC=arm-linux-gnueabihf-gcc --host=$TARGET --with-tlib=ncurses
    make all –j4 export CC=arm-linux-gnueabihf-gcc

The configure output is:

    ....
     checking for stack_t... yes
     checking whether stack_t has an ss_base field... no
    checking --with-tlib argument... ncurses
    checking for linking with ncurses library... OK
    checking for tgetent()... yes
    checking whether we talk terminfo... yes
    checking what tgetent() returns for an unknown terminal... zero
    checking whether termcap.h contains ospeed... yes
    checking whether termcap.h contains UP, BC and PC... yes
    checking whether tputs() uses outfuntype... no
    checking whether sys/select.h and sys/time.h may both be included... yes
    ...

However I get now a linking error:

    …
    link.sh: $LINK_AS_NEEDED set to 'yes': invoking linker directly.
    arm-linux-gnueabihf-gcc   -L/usr/local/lib -Wl,--as-needed    -o vim objects/arabic.o objects/buffer.o objects/blowfish.o objects/crypt.o objects/crypt_zip.o objects/dict.o objects/diff.o objects/digraph.o objects/edit.o objects/eval.o objects/evalfunc.o objects/ex_cmds.o objects/ex_cmds2.o objects/ex_docmd.o objects/ex_eval.o objects/ex_getln.o objects/farsi.o objects/fileio.o objects/fold.o objects/getchar.o objects/hardcopy.o objects/hashtab.o  objects/if_cscope.o objects/if_xcmdsrv.o objects/list.o objects/mark.o objects/memline.o objects/menu.o objects/misc1.o objects/misc2.o objects/move.o objects/mbyte.o objects/normal.o objects/ops.o objects/option.o objects/os_unix.o objects/pathdef.o objects/popupmnu.o objects/pty.o objects/quickfix.o objects/regexp.o objects/screen.o objects/search.o objects/sha256.o objects/spell.o objects/spellfile.o objects/syntax.o objects/tag.o objects/term.o objects/terminal.o objects/ui.o objects/undo.o objects/userfunc.o objects/version.o objects/window.o            objects/netbeans.o objects/channel.o  objects/charset.o objects/json.o objects/main.o objects/memfile.o objects/message.o    -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -lnsl  -lselinux -lncurses -ldl
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lSM
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lICE
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lXpm
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lXt
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lX11
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lXdmcp
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lSM
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lICE
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lselinux
    /usr/lib/gcc-cross/arm-linux-gnueabihf/5/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lncurses
    collect2: error: ld returned 1 exit status
    link.sh: Linking failed
    …

I tried export LDFLAGS=~/work/ncurses/install/lib for ncurses but the error -lncurses remains. Any suggestions? Thanks in advance

**Comment by k-takata** [comment]
`~` is not exactly as same as `$HOME`.
https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html

**Comment by nuko8** [comment]
@companion-vu 

> I tried export LDFLAGS=~/work/ncurses/install/lib 

And, you need to add `-L` between `=` and `~`.  Seems like you were making the same typo that I did.  I thought you already noticed @k-takata 's comment which correctly includes `-L`, so I didn't correct my typo to avoid duplication.  Sorry for having kept you confusing.

So, just like he is suggesting it, how about trying the following?
```
LDFLAGS=-L$HOME/work/ncurses/install/lib
```
Also, it appears that, to complete the cross compile, you need to have an X11 environment for armv7 on your Ubuntu box.  Although that is obviously related to the current issue, that's still another issue and could be laborious for its own sake.  If you find the latter the case,  I'd like to suggest that you run the configure script with `--without-x` ~~until this ncurses issue is resolved.~~ while this ncurses issues is not resolved.

**Comment by companion-vu** [comment]
@nuko8 Yes you are right!
@k-takata I am suprised that there is a difference. Sorry that I didn't get it immediately.
I started all over again:

    export TARGET=arm-linux-gnueabihf
    export vim_cv_toupper_broken="set"
    export vim_cv_terminfo="yes"
    export LDFLAGS=' -L$HOME/work/ncurses/install/lib/ -L$HOME/work/ncurses/install/bin/'
    export PREFIX=$HOME/work/vim/install/
    ../src/configure --target=$TARGET --prefix=$PREFIX --with-sysroot=$TARGET --with-arch=$TARGET CC=arm-linux-gnueabihf-gcc --host=$TARGET --with-tlib=ncurses --without-x

I get now the error: 

   ...
    checking for stack_t... yes
    checking whether stack_t has an ss_base field... no
    checking --with-tlib argument... ncurses
    checking for linking with ncurses library... OK
    checking for tgetent()... yes
    checking whether we talk terminfo... (cached) yes
    checking what tgetent() returns for an unknown terminal... configure: error: failed to compile test program.

I have 3 questions: 
1) How can I come around this error? I tried it with and without --without-x
2) I have now two parameters to set: vim_cv_toupper_broken and vim_cv_terminfo. For what are they good?
3) Which X11 package do I have to cross compile? Is xorg enough?
Thanks

**Comment by nuko8** [comment]
> I have 3 questions:
> 
> 1.  How can I come around this error? I tried it with and without --without-x
> 2.  I have now two parameters to set: vim_cv_toupper_broken and vim_cv_terminfo. For what are they good?
> 3.  Which X11 package do I have to cross compile? Is xorg enough?

That all depends...  Actually, those questions are making me wonder because they are rather things you need to know about and decide what to do with before starting cross compiling.  I thought you were asking just about the correct usage of `--with-tlib` for the cross compiling you had been doing on you machine, and I had replied to you as such.  All of a sudden, the questions sound like what you've been actually asking is how to do cross compiling in general.

For that, please refer to [src/INSTALLx.txt](https://github.com/vim/vim/blob/master/src/INSTALLx.txt) for Vim specific issues.  As to general instructions on how to cross compile, how about following the [pointer](http://www.kegel.com/crosstool/) mentioned in the document?  The article itself is designated as out-of-date, but the author kindly gives a couple of up-to-date references and their pointer at the head of the article.

Also, please note that this is the issue tracker for Vim.  That means, questions too general or relevant little to Vim often remain unanswered or are dismissed out of hand.  So be careful where/what to ask.

**Comment by loctran17** [comment]
I found the same issue when compile the vim as well. Here is my solution:
--with-tlib=ncurses should replace by --with-tlib=tinfo . libtinfo was generated from ncurses when you add the --with-termlib whilecompiling ncurse

**Comment by leon-anavi** [comment]
Hi,

I have experienced the same issue while building Vim 8.1 through Yocto/OE recipe as well as for Buildroot. The solution was to add: 
```
vim_cv_tgetent=zero
```
For details: http://lists.openembedded.org/pipermail/openembedded-devel/2018-July/119160.html

Best regards,
Leon

**Comment by xspeed1989** [comment]
> Hi,
> 
> I have experienced the same issue while building Vim 8.1 through Yocto/OE recipe as well as for Buildroot. The solution was to add:
> 
> ```
> vim_cv_tgetent=zero
> ```
> 
> For details: http://lists.openembedded.org/pipermail/openembedded-devel/2018-July/119160.html
> 
> Best regards,
> Leon

works for me, thanks!
